src/subspace_store.c

```diff
@@ -213,6 +215,7 @@ ts_subspace_store_get(SubspaceStore *store, Point *target)
 		vec = ((SubspaceStoreInternalNode *) match->storage)->vector;
 	}
 	Assert(match != NULL);
+	/* NOLINTNEXTLINE(clang-analyzer-core.NullDereference) */
 	return match->storage;
```

Comment:
Not sure why you need this: there is an assert on the line before, so it should not generate a warning.

Reply:
When I did this PR, it was in response to another change that made PR builds fail because the linter ran on release builds (Mac OS run) where assertions are not enabled. This made the code build without warnings and tests pass again.

However, [another change was committed](https://github.com/timescale/timescaledb/commit/417908f19b25194ab463b0b8edb73e2247c2ba74), seemingly without review, to add assertions to release builds in order to silence linter warnings. Personally, I don't think keeping assertions in release builds just to silence the linter is not the right thing to do.

Reply:
Assertions should not be enabled in release builds. See my other comment on this.

Reply:
Not sure I follow:
* The lint comment is only relevant for `clang-tidy` and it will see the assert as well, so one of them is not necessary.
* AIUI, there is no "release build" for `clang-tidy` (or shouldn't be).

Yes, we should not have asserts in the release code (even when we do a test build ourselves), but AFAICT, the situation you describe above is more related to the build setup rather than the code.

* Debug builds should run the linter and have assertions enabled.
* Release build do not need to run the linter (but can) and should not have assertions enabled.

Reply:
What you describe is not how things currently work. And, I am not convinced it should work as you suggest either. For instance, I see nothing wrong with running the linter on a release build since that fully represents the code that we ship.

Our CI currently runs the linter on the Mac OS build, and it was changed to a release build recently (to speed up the build processes). But since that also disabled assertions, another commit re-enabled assertions on the release build. Thus, we have no CI build that fully matches a "real" release build, which is a problem for many reasons; including, e.g., detecting variables that are only used in Assertions, and thus would fail a real release build.

This is the problem of inserting unreviewed commits in existing PRs. This PR attempts to fix the issues described above so that we can run a "real" release build without Assertions, including linter.

Reply:
I'm fine with adding these comments for this PR, but I see no reason to run `clang-tidy` on release builds and I think this will not work long-term.

* The only difference between the release build and the debug build is the use of the `NDEBUG`, which disables assertions, and different optimization levels. It is in theory possible to add any sort of code in a release build by looking at this flag, but I do not think that is the main problem.
* For working code, the difference between compiling with `-DNDEBUG=1` and not is only that some variables might end up as unused. This is caught using `-Wunused` for a normal compile (exists for both clang and gcc).
* Running `clang-tidy` does a more advanced checking of the AST and does not really compile the code. It does run the preprocessor though. Running `clang-tidy` makes the compile time go up *a lot* and not something we want to do unless necessary.
* Not counting the unused variable above (which is caught using `-Wunused`) we will not catch any additional errors with `-DNDEBUG=1` since `clang-tidy` does not compile the code (it just looks at the AST).
* Adding a comment for each assertion will be forgotten at some point.
* The comments are specific for `clang-tidy`.

Reply:
Sounds like you are arguing for not running `clang-tidy` at all, or do you mean something else? In any case, this PR doesn't turn on `clang-tidy`---it was running in the CI builds prior to this PR.

If we do agree that we should run `clang-tidy` in the CI system, then I don't get the argument around it being slow. It is what it is and it is probably faster to run it on the release build in that case, since it won't analyze the extra testing code. For your private builds you can do whatever you want, including disabling `clang-tidy`.

As you point out yourself, `clang-tidy` doesn't compile the code, so I don't understand what you are trying to say about various compiler optimizations and what it has to do with static analysis.